src/nix-channel/nix-channel.cc

```diff
@@ -168,7 +168,12 @@ static int main_nix_channel(int argc, char ** argv)
         nixDefExpr = settings.useXDGBaseDirectories ? createNixStateDir() + "/defexpr" : home + "/.nix-defexpr";
 
         // Figure out the name of the channels profile.
-        profile = profilesDir() + "/channels";
+        // For backwards-compatibility, install the root channels under
+        // a custom location, as these are also used as "global" channeles, and
+        // their location is hardcoded in a number of places.
+        profile = getuid() == 0
+            ? settings.nixStateDir + "/profiles/per-user/root/channels"
```

Comment:
```c++
settings.nixStateDir + "/profiles/per-user/root/channels"
```
occurs in a number of places (along with a similar `fmt(...)` equivalent). I would like to factor that out into a function.

Reply:
Frankly, we should do a
```bash
git grep /per-user
```
and do an audit + de-dup of all of that

Reply:
> Frankly, we should do a
> 
> ```shell
> git grep /per-user
> ```
> 
> and do an audit + de-dup of all of that, but this is separate from this PR / this issue.

Yeah, I've started doing that. Most seem to be legit/OK though.